### PR TITLE
feat(renderer-app): add Monaco app to whiteboard

### PR DESCRIPTION
**Pull request by Cheerego7**

### CODE REVIEW

**Review by BlackHole1 — CHANGES_REQUESTED** [review]
fix renderer-app cannot find monaco-editor patch:

```diff
From 209a55abb10ba2b80cbf12539e778c8a7c53bae4 Mon Sep 17 00:00:00 2001
From: BlackHole1 <158blackhole@gmail.com>
Date: Sun, 26 Sep 2021 13:58:06 +0800
Subject: [PATCH] fix(renderer-app): cannot find monaco-editor path

---
 desktop/renderer-app/package.json             |  1 +
 .../src/tasks/init-register-apps.ts           | 20 ++++++++++++-----
 desktop/renderer-app/webpack/paths.js         |  1 +
 desktop/renderer-app/webpack/webpack.prod.js  | 19 +++++++++++++++-
 yarn.lock                                     | 22 ++++++++++++++++++-
 5 files changed, 56 insertions(+), 7 deletions(-)

diff --git a/desktop/renderer-app/package.json b/desktop/renderer-app/package.json
index e32befa..6ee1c39 100644
--- a/desktop/renderer-app/package.json
+++ b/desktop/renderer-app/package.json
@@ -64,6 +64,7 @@
     "@typescript-eslint/eslint-plugin": "^4.29.1",
     "@typescript-eslint/parser": "^4.29.1",
     "agora-electron-sdk": "^3.4.7-rc.1-build.817",
+    "copy-webpack-plugin": "^9.0.1",
     "cross-env": "^7.0.3",
     "css-loader": "^5.2.4",
     "dotenv-flow": "^3.2.0",
diff --git a/desktop/renderer-app/src/tasks/init-register-apps.ts b/desktop/renderer-app/src/tasks/init-register-apps.ts
index a43380e..dc7f789 100644
--- a/desktop/renderer-app/src/tasks/init-register-apps.ts
+++ b/desktop/renderer-app/src/tasks/init-register-apps.ts
@@ -1,16 +1,13 @@
 import { WindowManager } from "@netless/window-manager";
-import { pathToFileURL } from "url";
+import path from "path";
 
 const registerApps = async (): Promise<void> => {
-    const urlToLoader = pathToFileURL(require.resolve("monaco-editor/min/vs/loader")).toString();
-    const urlToVS = urlToLoader.replace(/\/loader\.js$/, "");
-
     await WindowManager.register({
         kind: "Monaco",
         appOptions: {
             loader: {
                 paths: {
-                    vs: urlToVS,
+                    vs: `file://${getVSPath()}`,
                 },
             },
         },
@@ -24,3 +21,16 @@ const registerApps = async (): Promise<void> => {
 export const initRegisterApps = async (): Promise<void> => {
     await registerApps();
 };
+
+const getVSPath = (): string => {
+    if (process.env.NODE_ENV === "production") {
+        return path.join(__dirname, "static", "monaco-editor", "min", "vs");
+    }
+
+    const nodeModulesName = "node_modules";
+    const nodeModulesEndIndexByPath = __dirname.indexOf(nodeModulesName) + nodeModulesName.length;
+
+    const rootNodeModulesPath = __dirname.slice(0, nodeModulesEndIndexByPath);
+
+    return path.join(rootNodeModulesPath, "monaco-editor", "min", "vs");
+};
diff --git a/desktop/renderer-app/webpack/paths.js b/desktop/renderer-app/webpack/paths.js
index a23827e..20d79db 100644
--- a/desktop/renderer-app/webpack/paths.js
+++ b/desktop/renderer-app/webpack/paths.js
@@ -11,6 +11,7 @@ module.exports = {
     appBuild: resolvePath("dist"),
     appHtml: resolvePath("public", "index.html"),
     appNodeModules: resolvePath("node_modules"),
+    rootNodeModules: resolvePath("..", "..", "node_modules"),
     entryFile: resolvePath("src", "index.ts"),
     tsConfig: resolvePath("tsconfig.json"),
     envConfig: resolvePath("..", "..", "config"),
diff --git a/desktop/renderer-app/webpack/webpack.prod.js b/desktop/renderer-app/webpack/webpack.prod.js
index 8f37b8c..6de1166 100644
--- a/desktop/renderer-app/webpack/webpack.prod.js
+++ b/desktop/renderer-app/webpack/webpack.prod.js
@@ -1,12 +1,29 @@
+const path = require("path");
 const paths = require("./paths");
 const { merge } = require("webpack-merge");
 const common = require("./webpack.common.js");
+const CopyPlugin = require("copy-webpack-plugin");
 const BundleAnalyzerPlugin = require("webpack-bundle-analyzer").BundleAnalyzerPlugin;
 
 module.exports = merge(common, {
     mode: "production",
 
-    plugins: [process.env.ANALYZER && new BundleAnalyzerPlugin()].filter(Boolean),
+    plugins: [
+        new CopyPlugin({
+            patterns: [
+                {
+                    from: path.join(paths.rootNodeModules, "monaco-editor", "min", "vs"),
+                    to: path.join(paths.appBuild, "static", "monaco-editor", "min", "vs"),
+                },
+                {
+                    // inject monaco-editor source-map
+                    from: path.join(paths.rootNodeModules, "monaco-editor", "min-maps", "vs"),
+                    to: path.join(paths.appBuild, "static", "monaco-editor", "min-maps", "vs"),
+                },
+            ],
+        }),
+        process.env.ANALYZER && new BundleAnalyzerPlugin(),
+    ].filter(Boolean),
 
     output: {
         filename: "static/js/[name].[contenthash:8].js",
diff --git a/yarn.lock b/yarn.lock
index f8ee71f..d49584d 100644
--- a/yarn.lock
+++ b/yarn.lock
@@ -6119,6 +6119,19 @@ copy-to@^2.0.1:
   resolved "https://registry.yarnpkg.com/copy-to/-/copy-to-2.0.1.tgz#2680fbb8068a48d08656b6098092bdafc906f4a5"
   integrity sha1-JoD7uAaKSNCGVrYJgJK9r8kG9KU=
 
+copy-webpack-plugin@^9.0.1:
+  version "9.0.1"
+  resolved "https://registry.yarnpkg.com/copy-webpack-plugin/-/copy-webpack-plugin-9.0.1.tgz#b71d21991599f61a4ee00ba79087b8ba279bbb59"
+  integrity sha512-14gHKKdYIxF84jCEgPgYXCPpldbwpxxLbCmA7LReY7gvbaT555DgeBWBgBZM116tv/fO6RRJrsivBqRyRlukhw==
+  dependencies:
+    fast-glob "^3.2.5"
+    glob-parent "^6.0.0"
+    globby "^11.0.3"
+    normalize-path "^3.0.0"
+    p-limit "^3.1.0"
+    schema-utils "^3.0.0"
+    serialize-javascript "^6.0.0"
+
 core-js-compat@^3.14.0, core-js-compat@^3.16.0, core-js-compat@^3.8.1:
   version "3.16.1"
   resolved "https://registry.yarnpkg.com/core-js-compat/-/core-js-compat-3.16.1.tgz#c44b7caa2dcb94b673a98f27eee1c8312f55bc2d"
@@ -8023,7 +8036,7 @@ fast-glob@^2.2.6:
     merge2 "^1.2.3"
     micromatch "^3.1.10"
 
-fast-glob@^3.1.1:
+fast-glob@^3.1.1, fast-glob@^3.2.5:
   version "3.2.7"
   resolved "https://registry.yarnpkg.com/fast-glob/-/fast-glob-3.2.7.tgz#fd6cb7a2d7e9aa7a7846111e85a196d6b2f766a1"
   integrity sha512-rYGMRwip6lUMvYD3BTScMwT1HtAs2d71SMv66Vrxs0IekGZEjhM0pcMfjQPnknBt2zeCwQMEupiN02ZP4DiT1Q==
@@ -8739,6 +8752,13 @@ glob-parent@^5.1.2, glob-parent@~5.1.2:
   dependencies:
     is-glob "^4.0.1"
 
+glob-parent@^6.0.0:
+  version "6.0.1"
+  resolved "https://registry.yarnpkg.com/glob-parent/-/glob-parent-6.0.1.tgz#42054f685eb6a44e7a7d189a96efa40a54971aa7"
+  integrity sha512-kEVjS71mQazDBHKcsq4E9u/vUzaLcw1A8EtUeydawvIWQCJM0qQ08G1H7/XTjFUulla6XQiDOG6MXSaG0HDKog==
+  dependencies:
+    is-glob "^4.0.1"
+
 glob-promise@^3.4.0:
   version "3.4.0"
   resolved "https://registry.yarnpkg.com/glob-promise/-/glob-promise-3.4.0.tgz#b6b8f084504216f702dc2ce8c9bc9ac8866fdb20"
-- 
2.33.0
```

File: desktop/renderer-app/src/components/AppStoreButton/index.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    return (
+        <>
+            <Tooltip placement="right" key="apps" title="Apps">
+                <div className="apps-button" onClick={() => setAppStoreIsVisible(true)}>
+                    <svg focusable="false" width="1em" height="1em" viewBox="0 0 24 24">
+                        <path
+                            d="M16 20h4v-4h-4m0-2h4v-4h-4m-6-2h4V4h-4m6 4h4V4h-4m-6 10h4v-4h-4m-6 4h4v-4H4m0 10h4v-4H4m6 4h4v-4h-4M4 8h4V4H4v4z"
+                            fill="currentColor"
+                        ></path>
+                    </svg>
```
Comment (BlackHole1) [inline desktop/renderer-app/src/components/AppStoreButton/index.tsx]: use svg file

File: desktop/renderer-app/webpack/webpack.common.js [hunk 2]
```diff
@@ -122,6 +122,7 @@ module.exports = {
 
     optimization: {
         runtimeChunk: "single",
+        moduleIds: "named",
```
Comment (BlackHole1) [inline desktop/renderer-app/webpack/webpack.common.js]: Revoke

File: desktop/renderer-app/package.json [hunk 3]
```diff
@@ -29,6 +30,7 @@
     "mobx": "^6.1.0",
     "mobx-react": "^7.1.0",
     "mobx-react-lite": "^3.2.0",
+    "monaco-editor": "^0.27.0",
```
Comment (BlackHole1) [inline desktop/renderer-app/package.json]: move to `devDependencies`

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

**Comment by Cheerego7** [comment]
Thank you @BlackHole1  for your assistance

**Review by BlackHole1 — APPROVED** [review]

**Pull request merged**